### PR TITLE
Adding ParsedCardinality

### PR DESCRIPTION
**Pull request by cbuescher**
First stab at adding parsing one of the single value aggregations from xContent.
This is WIP against our current feature branch.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java [hunk 1]
```diff
@@ -63,11 +66,12 @@ public final String getName() {
 
     //TODO the only way to avoid duplicating this method is making Aggregation extend ToXContent
     //and declare toXContent as a default method in it. Doesn't sound like the right thing to do.
+    @Override
     public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
         //TODO move TYPED_KEYS_DELIMITER constant out of InternalAggregation
         // Concatenates the type and the name of the aggregation (ex: top_hits#foo)
         builder.startObject(String.join(InternalAggregation.TYPED_KEYS_DELIMITER, getType(), name));
-        if (metadata.isEmpty() == false) {
+        if (this.metadata != null) {
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java]: I changed this (and the way the class interacts with the internal field) because I wanted to get close to what InternalAggregation allows for metadata. I think it can be null there, in which case the method returns null and we don't write anything to xContent.

File: core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java [hunk 2]
```diff
@@ -36,21 +35,25 @@
 
     //TODO move CommonFields out of InternalAggregation
     protected static void declareCommonFields(ObjectParser<? extends ParsedAggregation, Void> objectParser) {
-        objectParser.declareObject((parsedAgg, metadata) -> parsedAgg.metadata.putAll(metadata),
+        objectParser.declareObject((parsedAgg, metadata) -> parsedAgg.metadata = metadata,
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java]: I pushed a very similar change to our branch. There was a discrepancy between empty and null handling in this class compared to InternalAggregation, I aligned it to InternalAggregation. this should not be needed anymore here.

File: core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java (hunk 1 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java]: as I said above, that makes sense, I adjusted it on our branch that way.

File: core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java [hunk 3]
```diff
@@ -36,21 +35,25 @@
...
                 (parser, context) -> parser.map(), InternalAggregation.CommonFields.META);
     }
 
-    String name;
-    final Map<String, Object> metadata = new HashMap<>();
+    private String name;
+    protected Map<String, Object> metadata;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java]: why protected? subclasses can use the getter to retrieve it? but it only gets set while parsing? was this change required?

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java [hunk 4]
```diff
@@ -73,6 +87,41 @@ protected void assertReduced(InternalCardinality reduced, List<InternalCardinali
         }
     }
 
+    public void testFromXContent() throws IOException {
+        InternalCardinality cardinality = createTestInstance();
+        String type = cardinality.getWriteableName();
+        String name = cardinality.getName();
+        ToXContent.Params params = new ToXContent.MapParams(Collections.singletonMap(RestSearchAction.TYPED_KEYS_PARAM, "true"));
+        boolean humanReadable = randomBoolean();
+        XContentType xContentType = XContentType.JSON; //randomFrom(XContentType.values());
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: I think you forgot to put back the randomization here

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java [hunk 5]
```diff
@@ -73,6 +87,41 @@ protected void assertReduced(InternalCardinality reduced, List<InternalCardinali
...
+        BytesReference originalBytes = toXContent(cardinality, xContentType, params, humanReadable);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: do we want to shuffle these fields?

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java [hunk 6]
```diff
@@ -73,6 +87,41 @@ protected void assertReduced(InternalCardinality reduced, List<InternalCardinali
...
+            assertEquals(XContentParser.Token.FIELD_NAME, parser.nextToken());
+            assertEquals(type + "#" + name, parser.currentName());
+            assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+            parsed = (ParsedCardinality)  parser.namedObject(Aggregation.class, type, name);
+            assertEquals(XContentParser.Token.END_OBJECT, parser.currentToken());
+            assertEquals(XContentParser.Token.END_OBJECT, parser.nextToken());
+            assertNull(parser.nextToken());
+        }
+        assertEquals(cardinality.getName(), parsed.getName());
+        assertEquals(cardinality.getValue(), parsed.getValue(), Double.MIN_VALUE);
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: what is Double.MIN_VALUE doing here?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/ParsedCardinality.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import java.io.IOException;
+
+public class ParsedCardinality extends ParsedAggregation implements Cardinality {
+
+    private long cardinalityValue;
+
+    @Override
+    public String getValueAsString() {
+        // for some reason, InternalCardinality formats the double value
+        return DocValueFormat.RAW.format((double) cardinalityValue);
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/ParsedCardinality.java]: This is weird as this value is not returned with the response. But we can hardcode the formatter because this agg doesn't support format? I am a bit confused. Ideally we wouldn't have to implement this at all.

**Review by javanna — APPROVED** [review]
left a few comments, LGTM otherwise

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java (hunk 5 above)
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: Maybe not always, but that would be nice indeed

**Review by tlrx — APPROVED** [review]
I second @javanna in all his comments and I don't have much to add. I'd love to have a common test in InternalAggregationTestCase for all aggregations, but I'm sure we'll get there soon.

File: core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java (hunk 1 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java]: okay, will rebase this then

File: core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java (hunk 2 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java]: okay, will rebase this then

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java (hunk 5 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: will add this

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java (hunk 6 above)
Comment (cbuescher) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: Thats just the expected delta between expeted/actual value for double comparissons in assertEquals()

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/ParsedCardinality.java (hunk 7 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/ParsedCardinality.java]: Yes, InternalCardinality doesn't print "value_as_string", but you can get a formated value using the java API using `valueAsString` defined in InternalNumericMetricsAggregation. That one uses the default formater, which is DocValueFormat.RAW. I don't think this can be reset in InternalCardinality, thats why I use it here to get the same behaviour as InternalCardinality.

**Comment by cbuescher** [comment]
@javanna @tlrx thanks a lot, I updated the PR. Let me know if I'm missing anything or if I should merge to the feature branch.

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/ParsedCardinality.java (hunk 7 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/ParsedCardinality.java]: ok can you expand your comment on why we do this and why it's ok? I also wish one day we get rid of it. I thought about throwing UnsupportedOperationException for a sec, but let's keep it for now as-is.

File: core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java [hunk 8]
```diff
@@ -46,6 +46,10 @@ public final String getName() {
         return name;
     }
 
+    protected void setName(String name) {
+        this.name = name;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java]: I think that with this protected setter, you can now make the instance member private?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/ParsedCardinality.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    public double value() {
+        return getValue();
+    }
+
+    @Override
+    public long getValue() {
+        return cardinalityValue;
+    }
+
+    private void setValue(long cardinalityValue) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/ParsedCardinality.java]: It is a bit pointless to have a private setter for a private field, but we need it for object parser I guess?

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java (hunk 6 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: ok

**Review by javanna — APPROVED** [review]
LGTM thanks @cbuescher . I wonder if we want to remove ParsedAggregationTests at this point, not as part of this PR though.

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/ParsedCardinality.java (hunk 9 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/ParsedCardinality.java]: yes, thats the idea

File: core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java (hunk 8 above)
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/ParsedAggregation.java]: done
